File: ios/PreisvergleichApp/Models/Product.swift
import Foundation

// Structures für API Responses
struct SearchResponse: Codable {
    let query: String
    let results: [ProductResult]
    let postalCode: String
    let totalResults: Int
    let searchTimeMs: Int
    
    enum CodingKeys: String, CodingKey {
        case query, results
        case postalCode = "postal_code"
        case totalResults = "total_results"
        case searchTimeMs = "search_time_ms"
    }
}

struct ProductResult: Codable, Identifiable {
    let id = UUID()
    let name: String
    let price: String
    let store: String
    let storeLogoUrl: String?
    let productUrl: String?
    let imageUrl: String?
    let availability: String
    let pricePerUnit: String?
    let unit: String?
    let offerValidUntil: String?
    
    private enum CodingKeys: String, CodingKey {
        case name, price, store, availability
        case storeLogoUrl = "store_logo_url"
        case productUrl = "product_url"
        case imageUrl = "image_url"
        case pricePerUnit = "price_per_unit"
        case unit
        case offerValidUntil = "offer_valid_until"
    }
}

struct Store: Codable {
    let storeId: String
    let name: String
    let logoUrl: String?
    let websiteUrl: String?
    let category: String
    
    enum CodingKeys: String, CodingKey {
        case storeId = "id"
        case name
        case logoUrl = "logo_url"
        case websiteUrl = "website_url"
        case category
    }
}

// Saved Search Model
struct SavedSearch: Codable, Identifiable {
    let id: String
    let name: String
    let query: String
    let postalCode: String
    let selectedStores: [String]
    let unit: String?
    let maxPrice: Double?
    
    init(name: String, query: String, postalCode: String, selectedStores: [String] = [], unit: String? = nil, maxPrice: Double? = nil) {
        self.id = UUID().uuidString
        self.name = name
        self.query = query
        self.postalCode = postalCode
        self.selectedStores = selectedStores
        self.unit = unit
        self.maxPrice = maxPrice
    }
} 

File: android/app/src/main/java/com/preisvergleich/android/data/model/Product.kt
package com.preisvergleich.android.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.math.BigDecimal

@Serializable
data class SearchRequest(
    val query: String,
    @SerialName("postal_code")
    val postalCode: String,
    @SerialName("selected_stores")
    val selectedStores: List<String>? = null,
    val unit: String? = null,
    @SerialName("max_price")
    @Serializable(with = BigDecimalSerializer::class)
    val maxPrice: BigDecimal? = null
)

@Serializable
data class ProductResult(
    val name: String,
    @Serializable(with = BigDecimalSerializer::class)
    val price: BigDecimal,
    val store: String,
    @SerialName("store_logo_url")
    val storeLogoUrl: String? = null,
    @SerialName("product_url")
    val productUrl: String? = null,
    @SerialName("image_url")
    val imageUrl: String? = null,
    val availability: String = "verfügbar",
    val unit: String? = null,
    @SerialName("price_per_unit")
    @Serializable(with = BigDecimalSerializer::class)
    val pricePerUnit: BigDecimal? = null
)

@Serializable
data class SearchResponse(
    val results: List<ProductResult>,
    val query: String,
    @SerialName("postal_code")
    val postalCode: String,
    @SerialName("total_results")
    val totalResults: Int,
    @SerialName("search_time_ms")
    val searchTimeMs: Int
)

@Serializable
data class Store(
    val id: String,
    val name: String,
    @SerialName("logo_url")
    val logoUrl: String? = null,
    @SerialName("website_url")
    val websiteUrl: String? = null,
    val category: String
)

@Serializable
data class StoresResponse(
    val stores: List<Store>
) 